_posts/2022-01-10-reassign-partitions.md

```diff
@@ -0,0 +1,539 @@
+```
+
+Wait till the pod gets into the `Ready` state.
+Once it is ready, we're all set to work through the examples.
+
+## Example 1: Decreasing replication factor
+
+When we just created the `my-topic-two` topic we specified `spec.replicas` as 4.
+In hindsight that's actually more replicas than we need, so let's reduce it to 3.
+To do this we can simply remove the last replica from the list of replicas of each partition of that topic.
+
+Since we're operating on a single topic it's going to be easiest to use the `--generate` mode of `kafka-reassign-partitions.sh` and then edit that (either by hand or using a tool like [jq](https://stedolan.github.io/jq/)).
+
+### Creating a proposal `reassignment.json` file
+
+As we discussed above, this file will have the topics that we need to reassign.
+
+Let's create our `topics.json` file in a file in the current directory of our local machine:
+```json
+{
+  "version": 1,
+  "topics": [
+    { "topic": "my-topic-two"}
+  ]
+}
+```
+
+After creating this file, copy it to the interactive pod container since we will be running all of our commands from there:
+
+```sh
+kubectl cp topics.json my-pod:/tmp/topics.json
+```
+
+We can now start a shell process in our interactive pod container and run the command inside it to generate our proposal `reassignment.json` data.
+Let's start the shell process:
+
+```sh
+kubectl exec -ti my-pod /bin/bash
+```
+
+Now it's time to generate the `reassignment.json` file. 
+We will use the `kafka-reassign-partitions.sh` command to generate our proposal `reassignment.json` data:
+
+```sh
+kubectl exec -ti my-pod /bin/bash -- bin/kafka-reassign-partitions.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 \
+--topics-to-move-json-file /tmp/topics.json \
+--broker-list 0,1,2,3 \
+--generate
+```
+Here `topics-to-move-json-file` points towards the `topics.json file` and `--broker-list` is the list of brokers we want to move our partitions onto.
+
+Once you run this command, you will be able to see the JSON data which is generated by the Kafka reassignment partition tool. You get the current replica assignment and the proposed `reassignment.json` data.
+
+```json
+Current partition replica assignment
+{"version":1,"partitions":[{"topic":"my-topic-two","partition":0,"replicas":[3,4,2,0],"log_dirs":["any","any","any","any"]},{"topic":"my-topic-two","partition":1,"replicas":[0,2,3,1],"log_dirs":["any","any","any","any"]},{"topic":"my-topic-two","partition":2,"replicas":[1,3,0,4],"log_dirs":["any","any","any","any"]}]}
+
+Proposed partition reassignment configuration
+{"version":1,"partitions":[{"topic":"my-topic-two","partition":0,"replicas":[0,1,2,3],"log_dirs":["any","any","any","any"]},{"topic":"my-topic-two","partition":1,"replicas":[1,2,3,4],"log_dirs":["any","any","any","any"]},{"topic":"my-topic-two","partition":2,"replicas":[2,3,4,0],"log_dirs":["any","any","any","any"]}]}
```

Comment:
How many brokers we have ... AFAIU we have 3, or? Why I see replicas broker ids even 3 and 4?